File: in_Java/0227 Basic Calculator II.java
class Solution {
    public int calculate(String s) {
        if(s == null || s.length() == 0) return 0;
        Stack<Integer> st = new Stack<>();
        int n = 0;
        char sign = '+';
        for(int i = 0; i < s.length(); i++){
            if(Character.isDigit(s.charAt(i))){
                n = n * 10 + s.charAt(i)-'0';
            }
            if(!Character.isDigit(s.charAt(i)) && s.charAt(i) != ' ' || i == s.length()-1){
                if(sign == '+') st.push(n);
                else if(sign == '-') st.push(-n);
                else if(sign == '*') st.push(st.pop() * n);
                else st.push(st.pop()/n);
                sign = s.charAt(i);
                n = 0;
            }
        }
        int res = 0;
        while(!st.isEmpty()) res += st.pop();
        return res;
    }
}

File: in_Python/0227 Basic Calculator II.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        s = s.strip()
        def check(x):
            return '0' <= x <= '9'
        L = []
        i = 0
        while i < len(s):
            j = i
            while j < len(s) and check(s[j]):
                j += 1
            L.append(s[i:j])
            while j < len(s) and s[j] == ' ':
                j += 1
            if j < len(s):
                L.append(s[j])
            i = j + 1
            while i < len(s) and s[i] == ' ':
                i += 1
        st = []
        i = 0
        #print(L)
        while i < len(L):
            x = L[i]
            if x not in ['/','*']:
                st.append(x)
                i += 1
            else:
                pre = int(st.pop())
                nex = int(L[i+1])
                if x == '/':
                    st.append(str(pre / nex))
                else:
                    st.append(str(pre * nex))
                i += 2
        #print(st)
        res = int(st[0])
        i = 1
        while i < len(st):
            if st[i] == '+':
                res += int(st[i+1])
            else:
                res -= int(st[i+1])
            i += 2
        return res
                
        